e2e: Add i18n editor tests

yuliyan opened the pull request:
#### Changes proposed in this Pull Request

- Update navbar `writeButton` selector to not use text string.
- Update gutenberg editor page `blockSearch` selector to not use text string.
- Add Change UI Language flow.
- Add new i18n editor spec that test if translated strings for various blocks from Core, Jetpack, Crowdsignal Forms are present on the page.


#### Testing instructions

- Run tests `cd test/e2e && LOCALE=fr yarn jest specs/specs-i18n/wp-i18n__editor.ts`. `LOCALE` can be `en`, `fr`, or `he`. Other locales will be skipped as we have only added translations for these three locales.

yuliyan committed:
e2e: Add i18n editor tests

yuliyan committed:
Rework some components and flows selectors w/o English texts.

yuliyan committed:
Add more block definitions.

yuliyan committed:
Fix change UI language flow not waiting for saving the settings

yuliyan committed:
Fix leaving page with unsaved changes. Fix opening child block settings panel. More blocks definitions.

yuliyan committed:
Add more core blocks

yuliyan committed:
Replace LoginFlow with LoginPage.

dpasque commented on packages/calypso-e2e/src/lib/pages/me/account-settings-page.ts:
```diff
@@ -52,4 +58,14 @@ export class AccountSettingsPage {
+	/**
+	 * Changes the UI language for the currently logged in user.
+	 */
+	async changeUILanguage( localeSlug: string ): Promise< void > {
```
For type safety, what do you think of making a union string type for local slug?

Something like:
```typescript
type LocaleSlug = 'en' | 'he' | 'fr';
```

I've found this increases safety a lot when specific strings are required!

yuliyan replied:
Updated to use the type def for `LanguageSlug` from `@automattic/languages` package.

dpasque commented on packages/calypso-e2e/src/lib/pages/me/account-settings-page.ts:
```diff
@@ -11,6 +11,12 @@ const selectors = {
+	// UI Language
+	uiLanguageButton: 'button.language-picker',
+	uiLanguageSearch: '.language-picker-component__search-desktop .search-component__input',
+	uiLanguageItem: '.language-picker-component__language-button',
```
Optional suggestion, but sometimes we'll make "dynamic" selectors where we have the selector take an argument to take that logic out of the calling function and keep it all in one place.

So for example, it would be replacing this line with:
```typescript
uiLanguageItem: (localeSlug: string) => `.language-picker-component__language-button:has([lang="${ localeSlug }"])`
```

Definitely optional, but I find it's nice to not have to try to splice selectors together later!

yuliyan replied:
Updated it.

dpasque commented on packages/calypso-e2e/src/lib/flows/change-ui-language-flow.ts:
```diff
@@ -0,0 +1,43 @@
+	 * then onto Account Settings.
+	 */
+	async changeUILanguage( localeSlug: string ): Promise< void > {
+		await new SidebarComponent( this.page ).waitForSidebarInitialization();
```
Just out of curiosity - was there anything specific you needed here to wait for in the sidebar? This may not be necessary!

yuliyan replied:
You are right, it turned out not to be needed. I've added it initially as I was working fixing a flaky step and it seemed like it fixed it, but it might've been just a fluke.

dpasque commented on packages/calypso-e2e/src/lib/flows/change-ui-language-flow.ts:
```diff
@@ -0,0 +1,43 @@
+
+		await Promise.all( [
+			this.page.waitForNavigation(),
+			this.page.click( 'a[href="/me/account"]' ),
```
Hm, i18n creates a unique set of challenges for some selectors - we usually rely on text, but that doesn't work here!

Having the ability to navigate this menu by href seems like it may be really useful for future tests - what do you about adding the ability to navigate the `me` menu by href to `me-sidebar-component.ts`?  https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/components/me/me-sidebar-component.ts

yuliyan replied:
Updated it to use either text label or href value.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -258,16 +258,16 @@ export class GutenbergEditorPage {
 	 * @returns {Promise<void>} No return value.
 	 */
-	async openSettings(): Promise< void > {
+	async openSettings( toggleSelector = selectors.settingsToggle ): Promise< void > {
```
An optional stylistic suggestion, but we've been trying to use [destructured objects for optional parameters](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/docs/style-guide-playwright.md#destructure-parameters) as much as possible - what do you think of that here?

worldomonation replied:
I think instead of passing in the selector as a param, we should update the selector stored in `selectors` object.

yuliyan replied:
I've updated the selector and removed the param.

dpasque commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -258,16 +258,16 @@ export class GutenbergEditorPage {
 	 *
```
I think it would be good to update the jsdoc with the new parameter, especially because it might confuse people who don't know that we need the parameter specifically for i18 considerations!

dpasque commented on test/e2e/specs/specs-i18n/wp-i18n__editor.ts:
```diff
@@ -0,0 +1,270 @@
+
+	describe.each( locales )( 'Editor translations (%s)', ( locale ) => {
+		it( 'Change UI language', async function () {
+			await page.goto( DataHelper.getCalypsoURL( '/' ) );
```
A quick warning: this route can be a bit dastardly in Playwright! It has caused us many woes haha.

The catch is that this route then usually triggers a redirect to `/home/<site>`. Playwright can outpace that redirect, then get tripped up when it happens!

As a result, it's usually best to wrap this route in a specific `waitForNavigation` as such:

```typescript
		await Promise.all( [
			page.waitForNavigation( { url: '**/home/**', waitUntil: 'load' } ),
		    page.goto( DataHelper.getCalypsoURL( '/' ) )
		] );
```

dpasque commented on test/e2e/specs/specs-i18n/wp-i18n__editor.ts:
```diff
@@ -0,0 +1,270 @@
+			'Translations for block: $blockName',
+			async ( block ) => {
+				const frame = await gutenbergEditorPage.getEditorFrame();
+				await frame.waitForTimeout( 2000 );
```
Anything in particular you're waiting for here? We try to avoid arbitrary time waits in favor of specific ones when possible - it tends to keep the scripts more stable over time!

yuliyan replied:
Just an oversight, it wasn't really needed.

dpasque commented on test/e2e/specs/specs-i18n/wp-i18n__editor.ts:
```diff
@@ -0,0 +1,270 @@
+
+				await frame.waitForSelector(
+					`.block-editor-block-card__title:has-text("${ block.blockPanelTitle }")`
+				);
```
Optional suggestion: but what do you thinking about breaking this out into multiple cases? I've found that having smaller steps can really speed up debugging - you can quickly see what specific step had failed, which can be harder to deduce from an error message that's like `couldn't find selector xyz` etc.

dpasque commented on test/e2e/specs/specs-i18n/wp-i18n__editor.ts:
```diff
@@ -0,0 +1,270 @@
+				await frame.click( block.blockEditorSelector );
+				if ( await frame.isVisible( '.block-editor-block-parent-selector__button' ) ) {
+					await frame.click( '.block-editor-block-parent-selector__button' );
+				}
```
Hmm, this area seems a little risky to me...

`isVisible` can be a deceitful method... It doesn't do any waiting, so it returns immediately. This can lead to race conditions where something becomes visible a couple milliseconds after this is checked (especially in the react world).

Alternatively: what do you think about making handlers in the block definitions at the top of the spec file? Like instead of having `blockEditorSelector`, you have a handler that takes the `frame` as a parameter, and does the selecting and clicking it needs. That way, blocks that have parent blocks can explicitly just do `frame.click` and rely on its autowaiting, while blocks that don't can just skip that part!

yuliyan replied:
I've reworked it slightly to race for either click on the block parent selector button or wait for `blockEditorSelector` to be `.is-selected`. That should make it a bit more reliable.

dpasque replied:
Ooo, clever! 👍

dpasque reviewed (commented):
Left some comments, many small and optional. Overall, nice job on this spec! I18n testing sure brings up some tricky challenges in E2E testing, and this spec tackles those well! Thanks for adding these tests! 😄 

Also, after this merges, would you be ready for the TeamCity build for I18N Playwright tests?

yuliyan committed:
Clean up changeUILanguage flow

yuliyan committed:
Use proper type for localeSlug param

yuliyan committed:
Remove select toggle selector for openSettings action.

yuliyan committed:
Use home route with waitForNavigation.

yuliyan committed:
Split block translations steps

yuliyan committed:
Improve selecting parent block flow

yuliyan commented:
@dpasque, thank you for the detailed review. I've addressed the issues and updated the PR.

> Also, after this merges, would you be ready for the TeamCity build for I18N Playwright tests?
I'm working on [migrating the existing i18n specs to Playwright](https://github.com/Automattic/wp-calypso/pull/57707) and I was wondering if we can update the TC script to run both Selenium and Playwright specs, or if it will be better to hold merging this PR until the existing i18n specs are migrated to Playwright?

yuliyan committed:
Add tsconfig reference to languages package.

dpasque reviewed (approved):
Changes LGTM! 🎉

dpasque commented:
@yuliyan 
>I'm working on migrating the existing i18n specs to Playwright and I was wondering if we can update the TC script to run both Selenium and Playwright specs, or if it will be better to hold merging this PR until the existing i18n specs are migrated to Playwright?

We can for sure have the TC scripts run both! Are you okay with it being two separate TC builds though that run side by side? Their setup ends up being different enough that it's been helpful so far to have them in different builds, but there's no problem having them both running together side by side as things transition. That's actually what we did with the PR tests!

yuliyan committed:
Add conditional describe blocks to skip tests for locale with missing translations.

yuliyan commented:
@dpasque, thank you for reviewing it again! Instead of looping through the locale translations within the spec, I've reworked it so that runs the tests only for the locale passed from env variable `LOCALE`. I wanted to keep it consistent with the other i18n specs and that's how they work.

> We can for sure have the TC scripts run both! Are you okay with it being two separate TC builds though that run side by side? Their setup ends up being different enough that it's been helpful so far to have them in different builds, but there's no problem having them both running together side by side as things transition. That's actually what we did with the PR tests!

Separate TC builds should be fine, I think. However, I've also finished working on [migrating the old i18n specs to Playwright](https://github.com/Automattic/wp-calypso/pull/57707), so we may consider merging them together and removing Selenium script right away.
